---
title: "Making miami plots from scratch"
author: "Julie White"
date: "May 26, 2020"
output:
  html_document:
    df_print: paged
---

This document acts as a tutorial for how to create a Miami plot from scratch, 
in case you wanted more control than that provided by the `miamiplot` R package. 

# Necessary packages and data
```{r}
library(dplyr)
library(ggplot2)
library(ggrepel)
library(patchwork)
```

For this tutorial, we will use a GWAS of the UK Biobank data, available from 
the Neale lab: http://www.nealelab.is/uk-biobank

Number of cigarettes currently smoked daily (current cigarette smokers)
```{r, results="hide", message=FALSE}
sourceurl <- "https://www.dropbox.com/s/2sr93zgr9j9kwga/3456.gwas.imputed_v3.both_sexes.tsv.bgz?dl=1"
tmp <- tempfile()

download.file(sourceurl,tmp, quiet = TRUE)

cigarettes <- data.table::fread(cmd = paste("gunzip -c", tmp), sep = "\t", 
                                header = TRUE, stringsAsFactors = FALSE, 
                                colClasses = c("character", "character", 
                                               "numeric", "numeric", 
                                               "character", rep("numeric", 7)), 
                                na.strings = "NaN", showProgress = FALSE)

rm(sourceurl, tmp)
```

The example data I've used doesn't have SNP name or gene annotation included.
But, the Neale lab does provide an annotation file that we can use.
```{r, results="hide", message=FALSE}
sourceurl <- "https://www.dropbox.com/s/puxks683vb0omeg/variants.tsv.bgz?dl=1"
tmp <- tempfile()

download.file(sourceurl,tmp, quiet = TRUE)

variantlist <- data.table::fread(cmd = paste("gunzip -c", tmp), sep = "\t", 
                                header = TRUE, stringsAsFactors = FALSE,
                                na.strings = "NaN", showProgress = FALSE)

rm(sourceurl, tmp)
```

Add relevant annotation information to the gwas data
```{r}
variant_columns <- c("variant", "chr", "pos", "rsid", "consequence",
                     "consequence_category", "info")

cigarettes <- merge(cigarettes, variantlist[, ..variant_columns], 
                    by = "variant")

rm(variant_columns)
```

# Preparing the plot dataframe, axis information, and labels
Convert chromosome column to numeric
```{r}
cigarettes$chr <- as.numeric(gsub(pattern = "X", replacement = 23, x = cigarettes$chr))
```

Compute the cumulative position of the SNP
```{r}
plot_df <- cigarettes %>% 
  group_by(chr) %>% 
  # Compute chromosome size
  summarise(chrlength = max(pos)) %>%  
  # Calculate cumulative position of each chromosome
  mutate(cumulativechrlength = cumsum(as.numeric(chrlength))-chrlength) %>% 
  select(-chrlength) %>%
  # Temporarily add the cumulative length of each chromosome to the initial 
  # dataset 
  left_join(cigarettes, ., by=c("chr"="chr")) %>%
  # Sort by chr then position 
  arrange(chr, pos) %>%
  # Add the position to the cumulative chromosome length to get the position of 
  # this probe relative to all other probes
  mutate(rel_pos = pos + cumulativechrlength) %>%
  # Calculate the logged p-value too
  mutate(logged_p = -log10(pval)) %>%
  select(-cumulativechrlength)
```

We do not want to display the cumulative position of SNP in bp, but just show the chromosome name instead, so need to make a dataframe saying where each chromosome is
```{r}
axis_df <- plot_df %>% 
  group_by(chr) %>% 
  summarize(chr_center=(max(rel_pos) + min(rel_pos)) / 2)
```

Calculate the maximum p-value so that we can control the limits of the plot
```{r}
maxp <- ceiling(max(plot_df$logged_p, na.rm = TRUE))
```

As an illustration, we'll label the top 5 SNPs from separate chromosomes on each
plot. 
```{r}
upper_labels<- plot_df %>%
  filter(beta > 0) %>%
  group_by(chr) %>%
  arrange(desc(logged_p)) %>%
  slice(1) %>%
  ungroup() %>%
  arrange(desc(logged_p)) %>%
  slice(1:5) %>%
  select(rsid)

lower_labels <- plot_df %>%
  filter(beta <= 0) %>%
  group_by(chr) %>%
  arrange(desc(logged_p)) %>%
  slice(1) %>%
  ungroup() %>%
  arrange(desc(logged_p)) %>%
  slice(1:5) %>%
  select(rsid)

label_list <- rbind(upper_labels, lower_labels)

plot_df <- plot_df %>%
  mutate(label = case_when(rsid %in% label_list$rsid ~ paste0(rsid, "\n", consequence_category), TRUE ~ ""))
```

# Plotting
Plot SNPs with positive beta values. I'm only plotting SNPs with p-values less 
than 0.005, to save time.
```{r}
upper_plot <- ggplot() + 
  geom_point(data = plot_df[which(plot_df$beta > 0 & plot_df$pval < 0.005),], 
             aes(x = rel_pos, y = logged_p, color = as.factor(chr)), 
             size = 0.25) +
  scale_color_manual(values = rep(c("black", "grey"), nrow(axis_df))) +
  scale_x_continuous(labels = axis_df$chr, 
                     breaks = axis_df$chr_center, 
                     expand = expansion(mult = 0.01)) +
  scale_y_continuous(limits = c(0, maxp), 
                     expand = expansion(mult = c(0.02, 0))) + 
  geom_hline(yintercept = -log10(1e-5), color = "red", linetype = "dashed", 
             size = 0.3) +
  geom_hline(yintercept = -log10(5e-8), color = "blue", linetype = "solid", 
             size = 0.3) +
  labs(x = "", y = bquote(atop('Positive beta values', '-log'[10]*'(p)'))) + 
  theme_classic() +
  theme(legend.position = "none", 
        axis.title.x = element_blank(),
        plot.margin = margin(b = 0, l = 10)) + 
  geom_label_repel(data = plot_df[which(plot_df$beta > 0 & 
                                          plot_df$pval < 0.005),],
                   aes(x = rel_pos, y = logged_p, label = label),
                   size = 2, segment.size = 0.2, point.padding = 0.3,
                   ylim = c(maxp / 3, NA), box.padding = 0.5)

upper_plot
```

Plot SNPs with negative beta values. The lower plot is mostly constructed the 
same way as the upper plot, but with the axis reversed.
```{r}
lower_plot <- ggplot() + 
  geom_point(data = plot_df[which(plot_df$beta <= 0 & plot_df$pval < 0.005),], 
             aes(x = rel_pos, y = logged_p, color = as.factor(chr)), 
             size = 0.25) +
  scale_color_manual(values = rep(c("black", "grey"), nrow(axis_df))) +
  scale_x_continuous(breaks = axis_df$chr_center, 
                     position = "top",
                     expand = expansion(mult = 0.01)) +
  scale_y_reverse(limits = c(maxp, 0), 
                     expand = expansion(mult = c(0, 0.02))) + 
  geom_hline(yintercept = -log10(1e-5), color = "red", linetype = "dashed", 
             size = 0.3) +
  geom_hline(yintercept = -log10(5e-8), color = "blue", linetype = "solid", 
             size = 0.3) +
  labs(x = "", y = bquote(atop('Negative beta values', '-log'[10]*'(p)'))) + 
  theme_classic() +
  theme(legend.position = "none", 
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        plot.margin = margin(t = 0, l = 10)) + 
  geom_label_repel(data = 
                     plot_df[which(plot_df$beta <= 0 & plot_df$pval < 0.005),],
                   aes(x = rel_pos, y = logged_p, label = label),
                   size = 2, segment.size = 0.2, point.padding = 0.3,
                   ylim = c(NA, -(maxp / 3)), box.padding = 0.5)

lower_plot
```

Add both together
```{r}
p <- upper_plot + lower_plot + plot_layout(ncol=1)
p
```

Save
```{r}
ggsave(filename = "MiamiPlot.png", plot = p, device = "png", width = 8, 
       height = 6.5, units = "in")
```

