---
title: "miamiplot"
author: "Julie White"
date: "` r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{miamiplot}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(miamiplot)
library(dplyr)
```

This document acts as a tutorial for using the miamiplot package to create Miami
plots for GWAS or EWAS results. Use `?FunctionName` in the R console to get the 
complete documentation of a given function.

# Data
For this tutorial, we will use two GWAS of the UK Biobank data, available from 
the Neale lab: http://www.nealelab.is/uk-biobank

Amount of alcohol drunk on a typical drinking day
```{r}
sourceurl <- "https://www.dropbox.com/s/58ssjbuxem71eik/20403.gwas.imputed_v3.both_sexes.tsv.bgz?dl=1"
tmp <- tempfile()

download.file(sourceurl,tmp)

alcoholdrinks <- data.table::fread(cmd = paste("gunzip -c", tmp), sep = "\t",
                                   header = TRUE, stringsAsFactors = FALSE,
                                   colClasses = c("character", "character", 
                                                  "numeric", "character",
                                                  rep("numeric", 7)),
                                   na.strings = "NaN")

rm(sourceurl, tmp)
```

Number of cigarettes currently smoked daily (current cigarette smokers)
```{r}
sourceurl <- "https://www.dropbox.com/s/2sr93zgr9j9kwga/3456.gwas.imputed_v3.both_sexes.tsv.bgz?dl=1"
tmp <- tempfile()

download.file(sourceurl,tmp)

cigarettes <- data.table::fread(cmd = paste("gunzip -c", tmp), sep = "\t", 
                                header = TRUE, stringsAsFactors = FALSE, 
                                colClasses = c("character", "character", 
                                               "numeric", "numeric", 
                                               "character", rep("numeric", 7)), 
                                na.strings = "NaN")

rm(sourceurl, tmp)
```

The example data I've used doesn't have SNP name or gene annotation included.
But, the Neale lab does provide an annotation file that we can use.
```{r}
sourceurl <- "https://www.dropbox.com/s/puxks683vb0omeg/variants.tsv.bgz?dl=1"
tmp <- tempfile()

download.file(sourceurl,tmp)

variantlist <- data.table::fread(cmd = paste("gunzip -c", tmp), sep = "\t", 
                                header = TRUE, stringsAsFactors = FALSE,
                                na.strings = "NaN")

rm(sourceurl, tmp)
```

Add relevant annotation information to the gwas data
```{r}
variant_columns <- c("variant", "chr", "pos", "rsid", "consequence",
                     "consequence_category", "info")

# Alcohol
alcoholdrinks <- merge(alcoholdrinks, variantlist[, ..variant_columns], 
                       by = "variant")


# Cigarettes
cigarettes <- merge(cigarettes, variantlist[, ..variant_columns], 
                    by = "variant")

rm(variant_columns)
```

# Plotting
Note: for the rest of the tutorial, I will be limiting the plots to the SNPs 
with p-values lower than 0.05, to speed up plotting time. 

## Assigning values to to the top and bottom sections of the plot
Say you wanted to plot the GWAS of cigarettes per day and wanted to separate
the plot by beta values, with the variants with positive beta values on top. 
Here, we're keeping the default y-axis labels. 
```{r}
ggmiami(data = cigarettes[which(cigarettes$pval < 0.05),], 
        split_by = "beta", split_at = 0, p = "pval")
```

Ah, this error indicates that we should check the class of our chromosome column.
```{r}
class(cigarettes$chr)
unique(cigarettes$chr)
```

And change it to numeric. Because "X" will not switch to numeric well, we will
also change that to 23
```{r}
cigarettes$chr <- as.numeric(gsub(pattern = "X", replacement = "23", 
                                  x = cigarettes$chr))

alcoholdrinks$chr <- as.numeric(gsub(pattern = "X", replacement = "23",
                                     x = alcoholdrinks$chr))
```

Double check..
```{r}
class(cigarettes$chr)
unique(cigarettes$chr)
```

Plot again...
```{r}
ggmiami(data = cigarettes[which(cigarettes$pval < 0.05),], 
        split_by = "beta", split_at = 0, p = "pval")
```

This GWAS also has information on whether the variant is a low confidence 
variant. Let's plot those on the bottom and high confidence variants on top and
label the y-axis accordingly. 
```{r}
ggmiami(data = cigarettes[which(cigarettes$pval < 0.05),], 
        split_by = "low_confidence_variant", 
        split_at = "false", p = "pval", top_ylab = "High Confidence Variants",
        bottom_ylab = "Low Confidence Variants")
```

Say you wanted to combine these two GWAS and plot them simultaneously. 
```{r}
# First label each study so we can tell them apart later. 
alcoholdrinks$pheno <- "alcoholdrinks"
cigarettes$pheno <- "cigarettes"

# Combine the two using the common column names. 
common_cols <- intersect(colnames(alcoholdrinks), colnames(cigarettes))
gwas_combined <- rbind(alcoholdrinks[, ..common_cols], 
                       cigarettes[, ..common_cols])


head(gwas_combined)
```

Plot, with alcohol drinks in the top plot
```{r}
# Filtering this one by p < 0.005 since there are now double the values to plot.
ggmiami(data = gwas_combined[which(gwas_combined$pval < 0.005),], 
        split_by = "pheno", split_at = "alcoholdrinks", p = "pval", 
        top_ylab = "Amount of alcohol on\n typical drinking day",
        bottom_ylab = "Daily number of cigarettes")
```

## Changing the significance lines
The suggestive and genome-wide significance lines can either be turned off or 
given different p-values. In this example we'll turn off the suggestive line and
plot the genome-wide line at a very small value: 5 x 10^-15
```{r}
ggmiami(data = gwas_combined[which(gwas_combined$pval < 0.005),], 
        split_by = "pheno", split_at = "alcoholdrinks", p = "pval", 
        top_ylab = "Amount of alcohol on\n typical drinking day",
        bottom_ylab = "Daily number of cigarettes", suggestive_line = NULL, 
        genome_line = 5e-15)
```

## Changing colors
The default chromosome colors are alternating black and grey. If you'd prefer to
supply your own alternating colors, you can do so using color names: 
```{r}
ggmiami(data = gwas_combined[which(gwas_combined$pval < 0.005),], 
        split_by = "pheno", split_at = "alcoholdrinks", p = "pval", 
        top_ylab = "Amount of alcohol on\n typical drinking day",
        bottom_ylab = "Daily number of cigarettes", 
        chr_colors = c("blue", "orange"))
```

Or using hexidecimal names: 
```{r}
ggmiami(data = gwas_combined[which(gwas_combined$pval < 0.005),], 
        split_by = "pheno", split_at = "alcoholdrinks", p = "pval", 
        top_ylab = "Amount of alcohol on\n typical drinking day",
        bottom_ylab = "Daily number of cigarettes", 
        chr_colors = c("#d8b365", "#5ab4ac"))
```

You may be thinking that red and blue for the significance lines don't exactly 
go well with the colors chosen. Those can be changed too
```{r}
ggmiami(data = gwas_combined[which(gwas_combined$pval < 0.005),], 
        split_by = "pheno", split_at = "alcoholdrinks", p = "pval", 
        top_ylab = "Amount of alcohol on\n typical drinking day",
        bottom_ylab = "Daily number of cigarettes", 
        chr_colors = c("#d8b365", "#5ab4ac"), suggestive_line_color = "black",
        genome_line_color = "#01665e")
```

You can also specify a list of colors equal to the number of chromosomes you'd 
like to plot. 
Here, using \code{brewer.pal}
```{r}
mycolors <- colorRampPalette(RColorBrewer::brewer.pal(12, "Paired"))(23)

ggmiami(data = gwas_combined[which(gwas_combined$pval < 0.005),], 
        split_by = "pheno", split_at = "alcoholdrinks", p = "pval", 
        top_ylab = "Amount of alcohol on\n typical drinking day",
        bottom_ylab = "Daily number of cigarettes", 
        chr_colors = mycolors, genome_line = NULL, suggestive_line = NULL)
```

And here using the ucsc genome browser colors, from \code{ggsci}
```{r}
mycolors <- ggsci::pal_ucscgb(palette = c("default"))(23)

ggmiami(data = gwas_combined[which(gwas_combined$pval < 0.005),], 
        split_by = "pheno", split_at = "alcoholdrinks", p = "pval", 
        top_ylab = "Amount of alcohol on\n typical drinking day",
        bottom_ylab = "Daily number of cigarettes", chr_colors = mycolors,
        genome_line = NULL, suggestive_line = NULL)
```

## Adding labels
To simply label the top 5 hits on each plot, all you need to do is supply the 
name of the column from which we should draw the lables. Here, we'll label the 
top 5 with the rsid.
```{r}
ggmiami(data = gwas_combined[which(gwas_combined$pval < 0.005),], 
        split_by = "pheno", split_at = "alcoholdrinks", p = "pval", 
        top_ylab = "Amount of alcohol on\n typical drinking day",
        bottom_ylab = "Daily number of cigarettes", hits_label_col = "rsid")
```

You can also label based on two columns, and change the number of labels: 
```{r}
ggmiami(data = gwas_combined[which(gwas_combined$pval < 0.005),], 
        split_by = "pheno", split_at = "alcoholdrinks", p = "pval", 
        top_ylab = "Amount of alcohol on\n typical drinking day",
        bottom_ylab = "Daily number of cigarettes", 
        hits_label_col = c("rsid", "consequence_category"), top_n_hits = 10)
```

If you would instead like to supply a list of which items to label, you can 
do that using \code{hits_label} to supply the list, and \code{hits_label_col} to
specify where these values will be found.
```{r}
ggmiami(data = gwas_combined[which(gwas_combined$pval < 0.005),], 
        split_by = "pheno", split_at = "alcoholdrinks", p = "pval", 
        top_ylab = "Amount of alcohol on\n typical drinking day",
        bottom_ylab = "Daily number of cigarettes", 
        hits_label = c("missense_variant", "stop_lost"),
        hits_label_col = "consequence")
```
This example produces a warning about special characters, in this case the "_".
We can ignore the warning in this case, becase I have provided items with exact
matches in \code{hits_label}.

Say we wanted to label the hits from each chromosome passing genome-wide 
significance. To do this, a little-front end work is required to identify these.
```{r}
gw_per_chromosome <- gwas_combined %>%
  filter(pval <= 5e-8) %>%
  group_by(pheno, chr) %>%
  arrange(desc(-log10(pval))) %>%
  slice(1)
```

```{r}
ggmiami(data = gwas_combined, 
        split_by = "pheno", split_at = "alcoholdrinks", p = "pval", 
        top_ylab = "Amount of alcohol on\n typical drinking day",
        bottom_ylab = "Daily number of cigarettes", 
        hits_label = gw_per_chromosome$rsid,
        hits_label_col = "rsid", top_n_hits = NULL)
```

Now for the plot.
```{r}
ggmiami(data = gwas_combined[which(gwas_combined$pval < 0.005),], 
        split_by = "pheno", split_at = "alcoholdrinks", p = "pval", 
        top_ylab = "Amount of alcohol on\n typical drinking day",
        bottom_ylab = "Daily number of cigarettes", 
        hits_label = gw_per_chromosome$rsid,
        hits_label_col = "rsid", top_n_hits = 25)
```




